File: apps/ios/verifd/Network/RetryPolicy.swift
import Foundation

/**
 * Retry policy with exponential backoff and jitter for network requests
 */
class RetryPolicy {
    private let maxAttempts: Int
    private let baseDelayMs: Int
    private let maxDelayMs: Int
    private let jitterFactor: Double
    
    init(maxAttempts: Int = 5, baseDelayMs: Int = 500, maxDelayMs: Int = 8000, jitterFactor: Double = 0.2) {
        self.maxAttempts = maxAttempts
        self.baseDelayMs = baseDelayMs
        self.maxDelayMs = maxDelayMs
        self.jitterFactor = jitterFactor
    }
    
    /**
     * Execute a network operation with retry logic
     */
    func execute<T>(_ operation: @escaping (Int) async throws -> T) async throws -> T {
        var lastError: Error?
        
        for attempt in 1...maxAttempts {
            do {
                return try await operation(attempt)
            } catch let error {
                lastError = error
                
                // Check if retryable
                if !isRetryable(error) || attempt == maxAttempts {
                    throw error
                }
                
                // Calculate delay with exponential backoff and jitter
                let delay = calculateDelay(attempt: attempt, error: error)
                try await Task.sleep(nanoseconds: UInt64(delay) * 1_000_000) // Convert ms to ns
            }
        }
        
        throw lastError ?? NetworkError.unknown
    }
    
    /**
     * Determine if an error is retryable
     */
    private func isRetryable(_ error: Error) -> Bool {
        if let networkError = error as? NetworkError {
            switch networkError {
            case .httpError(let statusCode, _):
                switch statusCode {
                case 429:  // Rate limited - retry with backoff
                    return true
                case 500...599:  // Server errors
                    return true
                case 400...499:  // Client errors (except 429)
                    return false
                default:
                    return false
                }
            case .timeout, .noConnection, .unknown:
                return true
            case .invalidResponse:
                return false
            }
        } else if error is URLError {
            let urlError = error as! URLError
            switch urlError.code {
            case .timedOut, .cannotFindHost, .cannotConnectToHost, .networkConnectionLost, .dnsLookupFailed:
                return true
            default:
                return false
            }
        }
        return false
    }
    
    /**
     * Calculate delay with exponential backoff and jitter
     */
    private func calculateDelay(attempt: Int, error: Error) -> Int {
        // Check for Retry-After header in case of rate limiting
        if let networkError = error as? NetworkError,
           case .httpError(_, let retryAfterSeconds) = networkError,
           let retryAfter = retryAfterSeconds {
            return retryAfter * 1000  // Convert seconds to milliseconds
        }
        
        // Exponential backoff: baseDelay * 2^(attempt-1)
        let exponentialDelay = baseDelayMs * (1 << (attempt - 1))
        
        // Cap at max delay
        let cappedDelay = min(exponentialDelay, maxDelayMs)
        
        // Add jitter to prevent thundering herd
        let jitter = Int(Double(cappedDelay) * jitterFactor * Double.random(in: 0...1))
        
        return cappedDelay + jitter
    }
}

/**
 * Custom error for network operations
 */
enum NetworkError: Error {
    case httpError(statusCode: Int, retryAfterSeconds: Int?)
    case timeout
    case noConnection
    case invalidResponse
    case unknown
    
    static func fromHTTPResponse(_ response: HTTPURLResponse) -> NetworkError {
        let retryAfter = response.value(forHTTPHeaderField: "Retry-After")
            .flatMap { Int($0) }
        
        return .httpError(statusCode: response.statusCode, retryAfterSeconds: retryAfter)
    }
}

File: apps/android/app/src/main/java/com/verifd/android/network/RetryPolicy.kt
package com.verifd.android.network

import kotlinx.coroutines.delay
import java.net.HttpURLConnection
import kotlin.math.min
import kotlin.random.Random

/**
 * Retry policy with exponential backoff and jitter for network requests
 */
class RetryPolicy(
    private val maxAttempts: Int = 5,
    private val baseDelayMs: Long = 500,
    private val maxDelayMs: Long = 8000,
    private val jitterFactor: Double = 0.2
) {
    /**
     * Execute a network operation with retry logic
     */
    suspend fun <T> execute(
        operation: suspend (attempt: Int) -> T
    ): T {
        var lastException: Exception? = null
        
        for (attempt in 1..maxAttempts) {
            try {
                return operation(attempt)
            } catch (e: Exception) {
                lastException = e
                
                // Check if retryable
                if (!isRetryable(e) || attempt == maxAttempts) {
                    throw e
                }
                
                // Calculate delay with exponential backoff and jitter
                val delay = calculateDelay(attempt, e)
                delay(delay)
            }
        }
        
        throw lastException ?: IllegalStateException("Retry failed")
    }
    
    /**
     * Determine if an exception is retryable
     */
    private fun isRetryable(exception: Exception): Boolean {
        return when (exception) {
            is NetworkException -> {
                when (exception.statusCode) {
                    // Don't retry client errors (except 429)
                    in 400..428 -> false
                    429 -> true  // Rate limited - retry with backoff
                    in 430..499 -> false
                    // Retry server errors
                    in 500..599 -> true
                    // Retry network errors
                    -1 -> true
                    else -> false
                }
            }
            is java.net.SocketTimeoutException,
            is java.net.UnknownHostException,
            is java.io.IOException -> true
            else -> false
        }
    }
    
    /**
     * Calculate delay with exponential backoff and jitter
     */
    private fun calculateDelay(attempt: Int, exception: Exception): Long {
        // Check for Retry-After header in case of rate limiting
        if (exception is NetworkException && exception.retryAfterSeconds != null) {
            return exception.retryAfterSeconds * 1000L
        }
        
        // Exponential backoff: baseDelay * 2^(attempt-1)
        val exponentialDelay = baseDelayMs * (1 shl (attempt - 1))
        
        // Cap at max delay
        val cappedDelay = min(exponentialDelay, maxDelayMs)
        
        // Add jitter to prevent thundering herd
        val jitter = (cappedDelay * jitterFactor * Random.nextDouble()).toLong()
        
        return cappedDelay + jitter
    }
}

/**
 * Custom exception for network errors with status code
 */
class NetworkException(
    val statusCode: Int,
    val retryAfterSeconds: Long? = null,
    message: String? = null,
    cause: Throwable? = null
) : Exception(message, cause) {
    companion object {
        fun fromHttpResponse(connection: HttpURLConnection): NetworkException {
            val retryAfter = connection.getHeaderField("Retry-After")?.toLongOrNull()
            return NetworkException(
                statusCode = connection.responseCode,
                retryAfterSeconds = retryAfter,
                message = "HTTP ${connection.responseCode}: ${connection.responseMessage}"
            )
        }
    }
}